picoquic/frames.c

```diff
@@ -904,32 +914,102 @@ uint8_t* picoquic_decode_stream_frame(picoquic_cnx_t* cnx, uint8_t* bytes, const
-            (stream->stop_sending_requested && !stream->stop_sending_sent)) {
-            /* Something can be sent */
-            /* if the stream is not active yet, verify that it fits under
-             * the max stream id limit */
-             /* Check parity */
-            if (IS_CLIENT_STREAM_ID(stream->stream_id) == cnx->client_mode) {
-                if (stream->stream_id <= cnx->max_stream_id_bidir_remote) {
+    for (int nb_pass = 0; nb_pass < 2; nb_pass++) {
+        stream = cnx->first_stream;
+
+        if (nb_pass == 0) {
+            /* Skip to the first non visited stream */
+            while (stream && stream->stream_id <= cnx->last_visited_stream_id) {
+                stream = stream->next_stream;
+            }
+        }
+
+        while (stream) {
+            if ((cnx->maxdata_remote > cnx->data_sent && stream->sent_offset < stream->maxdata_remote &&
+                (stream->is_active ||
+                (stream->send_queue != NULL && stream->send_queue->length > stream->send_queue->offset) ||
+                (stream->fin_requested && !stream->fin_sent))) ||
+                    (stream->reset_requested && !stream->reset_sent) ||
+                (stream->stop_sending_requested && !stream->stop_sending_sent)) {
+                /* Something can be sent */
+                /* if the stream is not active yet, verify that it fits under
+                 * the max stream id limit */
+                 /* Check parity */
+                if (IS_CLIENT_STREAM_ID(stream->stream_id) == cnx->client_mode) {
+                    if (stream->stream_id <= cnx->max_stream_id_bidir_remote) {
+                        break;
+                    }
+                }
+                else {
                     break;
                 }
             }
+
+            stream = stream->next_stream;
+
+            if (nb_pass > 0) {
+                /* Dont do the loop twice */
+                if (stream && stream->stream_id > cnx->last_visited_stream_id) {
+                    stream = NULL;
+                    break;
+                }
+            }
+        };
+
+        /* Do only one pass if we found a stream */
+        if (stream != NULL) {
+            break;
+        }
+    }
+
+    return stream;
+}
+
+typedef struct st_picoquic_stream_data_buffer_argument_t {
+    uint8_t * bytes;
+    size_t byte_index;
+    size_t space;
+    size_t length;
+    int is_fin;
+} picoquic_stream_data_buffer_argument_t;
+
+uint8_t * picoquic_provide_stream_data_buffer(void * context, size_t length, int is_fin)
+{
+    picoquic_stream_data_buffer_argument_t * data_ctx = (picoquic_stream_data_buffer_argument_t*)context;
+    uint8_t * buffer = NULL;
+
+    if (length <= data_ctx->space) {
+        data_ctx->length = length;
+
+        if (is_fin) {
+            data_ctx->is_fin = 1;
+            data_ctx->bytes[0] |= 1;
+        }
+
+        if (length < data_ctx->space) {
+            if (length == data_ctx->space - 1) {
```

Comment:
There are N bytes available, the application wants to write N-1 bytes. We can encode N bytes because then we don't need a length field, just a flag in the first byte. But if we had to encode "length=N-1", that would typically require 2 bytes, for a total of (2 + N-1)=N+1 bytes, larger than the packet size. We also don't want to avoid the length field, because the encoding would be shorter than the packet size, and other parts of the code might add a byte after that, e.g. padding,  which the receiver would mistake as data because of the "implicit length" encoding. So we work against that issue by inserting a single padding byte in front of the text header.